dex/networks/ltc/tx.go

```diff
@@ -0,0 +1,556 @@
+package ltc
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	"github.com/btcsuite/btcd/wire"
+
+	"lukechampine.com/blake3"
+)
+
+var byteOrder = binary.LittleEndian
+
+const (
+	pver               uint32 = 0                             // only protocol version 0 supported
+	maxTxInPerMessage         = wire.MaxMessagePayload/41 + 1 // wire.maxTxInPerMessage
+	maxTxOutPerMessage        = wire.MaxMessagePayload/       // wire.maxTxOutPerMessage
+		wire.MinTxOutPayload + 1
+	maxWitnessItemsPerInput = 500000 // from wire
+	maxWitnessItemSize      = 11000  // from wire
+)
+
+type decoder struct {
+	buf [8]byte
+	io.Reader
+	tee *bytes.Buffer
+}
+
+func newDecoder(r io.Reader) *decoder {
+	return &decoder{Reader: r}
+}
+
+func (d *decoder) mirror(b []byte) {
+	if d.tee != nil {
+		d.tee.Write(b)
+	}
+}
+
+func (d *decoder) readByte() (byte, error) {
+	b := d.buf[:1]
+	if _, err := io.ReadFull(d, b); err != nil {
+		return 0, err
+	}
+	d.mirror(b)
+	return b[0], nil
+}
+
+func (d *decoder) readUint16() (uint16, error) {
+	b := d.buf[:2]
+	if _, err := io.ReadFull(d, b); err != nil {
+		return 0, err
+	}
+	d.mirror(b)
+	return byteOrder.Uint16(b), nil
+}
+
+func (d *decoder) readUint32() (uint32, error) {
+	b := d.buf[:4]
+	if _, err := io.ReadFull(d, b); err != nil {
+		return 0, err
+	}
+	d.mirror(b)
+	return byteOrder.Uint32(b), nil
+}
+
+func (d *decoder) readUint64() (uint64, error) {
+	b := d.buf[:]
+	if _, err := io.ReadFull(d, b); err != nil {
+		return 0, err
+	}
+	d.mirror(b)
+	return byteOrder.Uint64(b), nil
+}
+
+// readOutPoint reads the next sequence of bytes from r as an OutPoint.
+func (d *decoder) readOutPoint(op *wire.OutPoint) error {
+	_, err := io.ReadFull(d, op.Hash[:])
+	if err != nil {
+		return err
+	}
+	d.mirror(op.Hash[:])
+
+	op.Index, err = d.readUint32()
+	return err
+}
+
+// wire.ReadVarInt a.k.a. CompactSize, not VARINT
+// https://en.bitcoin.it/wiki/Protocol_documentation#Variable_length_integer
+func (d *decoder) readCompactSize() (uint64, error) {
+	// Compact Size
+	// size <  253        -- 1 byte
+	// size <= USHRT_MAX  -- 3 bytes  (253 + 2 bytes)
+	// size <= UINT_MAX   -- 5 bytes  (254 + 4 bytes)
+	// size >  UINT_MAX   -- 9 bytes  (255 + 8 bytes)
+	chSize, err := d.readByte()
```

Comment:
Mainly for `mirror`/`tee`, but also to keep things a bit more local since it's a fairly straightforward encoding.  Will swing back to see.

I was even considering replicating what `wire.ReadVarBytes` does but more simply (see use in `readTxIn` and `readTxOut` that is followed by a `WriteVarBytes` if `tee` is set).